esp-hal/src/dma/mod.rs

```diff
@@ -732,58 +730,65 @@ where
+        let mut last_in_descr = unsafe { last_in_dscr_ptr.read_volatile() };
 
-        if dw0.owner() == Owner::Cpu && !dw0.is_empty() {
-            let descriptor_buffer = dw0.buffer;
-            let next_descriptor = dw0.next;
-
-            self.read_buffer_start = descriptor_buffer;
-            self.available = dw0.len();
-
-            dw0.set_owner(Owner::Dma);
-            dw0.set_length(0);
-            dw0.set_suc_eof(false);
+        if !last_in_descr.is_empty() && last_in_descr.owner() == Owner::Cpu {
+            let len = last_in_descr.len();
 
+            last_in_descr.set_owner(Owner::Dma);
+            last_in_descr.set_suc_eof(false);
             unsafe {
-                descr_address.write_volatile(dw0);
+                last_in_dscr_ptr.write_volatile(last_in_descr);
             }
 
-            self.last_seen_handled_descriptor_ptr = if next_descriptor.is_null() {
-                self.descriptors.as_mut_ptr()
-            } else {
-                next_descriptor
-            };
+            if self.last_seen_handled_descriptor_ptr.is_null() {
+                self.last_seen_handled_descriptor_ptr = last_in_dscr_ptr;
+                self.read_buffer_start = last_in_descr.buffer;
+                self.available = len;
+                return self.available;
+            }
+
+            if self.last_seen_handled_descriptor_ptr != last_in_dscr_ptr {
+                self.available += len;
+                return self.available;
+            }
         }
 
         self.available
     }
 
     fn pop(&mut self, data: &mut [u8]) -> Result<usize, DmaError> {
-        let avail = self.available;
+        let mut avail = self.available;
 
-        if avail < data.len() {
-            return Err(DmaError::Exhausted);
-        }
+        let mut remaining = data.len();
+        let mut offset = 0;
+        let mut descr_ptr = self.last_seen_handled_descriptor_ptr;
 
-        unsafe {
-            let dst = data.as_mut_ptr();
-            let src = self.read_buffer_start;
-            let count = self.available;
-            core::ptr::copy_nonoverlapping(src, dst, count);
+        while remaining != 0 && remaining >= avail {
+            unsafe {
+                let mut descr = descr_ptr.read_volatile();
+                let dst = data.as_mut_ptr().add(offset);
+                let src = descr.buffer;
+                let count = descr.len();
+                core::ptr::copy_nonoverlapping(src, dst, count);
+
+                descr.set_length(0);
+                descr_ptr.write_volatile(descr);
+
+                remaining -= count;
+                offset += count;
+                avail -= count;
+                descr_ptr = descr.next;
+            }
+
+            if descr_ptr.is_null() {
+                break;
+            }
         }
 
-        self.available = 0;
+        self.last_seen_handled_descriptor_ptr = core::ptr::null_mut();
+        self.available = avail;
         Ok(data.len())
```

Comment:
Should this be `Ok(offset)`? Otherwise I don't really understand why this method returns a `usize`.